src/nnet3/nnet-simple-component.cc

```diff
@@ -1011,6 +1022,196 @@ void PerElementScaleComponent::UnVectorize(
   scales_.CopyFromVec(params);
 }
 
+void PerElementOffsetComponent::Scale(BaseFloat scale) {
+  offsets_.Scale(scale);
+}
+
+void PerElementOffsetComponent::Resize(int32 dim) {
+  KALDI_ASSERT(dim > 0);
+  offsets_.Resize(dim);
+}
+
+void PerElementOffsetComponent::Add(BaseFloat alpha,
+                                   const Component &other_in) {
+  const PerElementOffsetComponent *other =
+      dynamic_cast<const PerElementOffsetComponent*>(&other_in);
+  KALDI_ASSERT(other != NULL);
+  offsets_.AddVec(alpha, other->offsets_);
+}
+
+PerElementOffsetComponent::PerElementOffsetComponent(
+    const PerElementOffsetComponent &component):
+    UpdatableComponent(component),
+    offsets_(component.offsets_) { }
+
+void PerElementOffsetComponent::SetZero(bool treat_as_gradient) {
+  if (treat_as_gradient) {
+    SetLearningRate(1.0);
+    is_gradient_ = true;
+  }
+  offsets_.SetZero();
+}
+
+void PerElementOffsetComponent::PerturbParams(BaseFloat stddev) {
+  CuVector<BaseFloat> temp_offsets(offsets_);
+  temp_offsets.SetRandn();
+  offsets_.AddVec(stddev, temp_offsets);
+}
+
+std::string PerElementOffsetComponent::Info() const {
+  std::stringstream stream;
+  BaseFloat offsets_mean = offsets_.Sum() / offsets_.Dim();
+  BaseFloat offsets_stddev = std::sqrt(VecVec(offsets_, offsets_) /
+                              offsets_.Dim()) - (offsets_mean * offsets_mean);
+  stream << Type() << ", input-dim=" << InputDim()
+         << ", output-dim=" << OutputDim()
+		 << ", offsets-mean=" << offsets_mean
+         << ", offsets-stddev=" << offsets_stddev
+         << ", learning-rate=" << LearningRate();
+  return stream.str();
+}
+
+Component* PerElementOffsetComponent::Copy() const {
+  PerElementOffsetComponent *ans = new PerElementOffsetComponent();
+  ans->learning_rate_ = learning_rate_;
+  ans->offsets_ = offsets_;
+  ans->is_gradient_ = is_gradient_;
+  return ans;
+}
+
+BaseFloat PerElementOffsetComponent::DotProduct(
+    const UpdatableComponent &other_in) const {
+  const PerElementOffsetComponent *other =
+      dynamic_cast<const PerElementOffsetComponent*>(&other_in);
+  return VecVec(offsets_, other->offsets_);
+}
+
+void PerElementOffsetComponent::Init(
+    BaseFloat learning_rate, int32 dim,
+    BaseFloat param_mean, BaseFloat param_stddev) {
+  UpdatableComponent::Init(learning_rate);
+  offsets_.Resize(dim);
+  KALDI_ASSERT(dim > 0 && param_stddev >= 0.0);
```

Comment:
Thanks-- then fix that too in your pull request.

On Mon, Nov 2, 2015 at 12:38 AM, hhadian <notifications@github.com> wrote:

> In src/nnet3/nnet-simple-component.cc
> <https://github.com/kaldi-asr/kaldi/pull/307#discussion_r43598253>:
>
> > +  return ans;
> > +}
> > +
> > +BaseFloat PerElementOffsetComponent::DotProduct(
> > +    const UpdatableComponent &other_in) const {
> > +  const PerElementOffsetComponent *other =
> > +      dynamic_cast<const PerElementOffsetComponent*>(&other_in);
> > +  return VecVec(offsets_, other->offsets_);
> > +}
> > +
> > +void PerElementOffsetComponent::Init(
> > +    BaseFloat learning_rate, int32 dim,
> > +    BaseFloat param_mean, BaseFloat param_stddev) {
> > +  UpdatableComponent::Init(learning_rate);
> > +  offsets_.Resize(dim);
> > +  KALDI_ASSERT(dim > 0 && param_stddev >= 0.0);
>
> Ok. This problem is also present in PerElementScaleComponent::Init.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/307/files#r43598253>.
>